### PR TITLE
Ct miss update actions list

### PR DESCRIPTION
**Pull request by ozshlomo**

### CODE REVIEW

File: lib/dpif-netdev.c [hunk 1]
```diff
@@ -6732,6 +6732,24 @@ smc_lookup_batch(struct dp_netdev_pmd_thread *pmd,
...
+    struct dp_netdev_actions* actions = NULL;
+
+    flow = mark_to_flow_find(pmd, mark);
+    if (flow)
+        actions = dp_netdev_flow_get_actions(flow);
+
+    netdev_flow_restore_state(port->netdev, mark, packet,
+                              (actions? actions->actions: NULL),
+                              (actions? actions->size : 0),
+                               &flow->offloaded_actions_len);
```
Comment (elibritstein) [inline lib/dpif-netdev.c]: lib/dpif-netdev.c: In function ‘process_partial_offload’:
lib/dpif-netdev.c:6750:37: error: ‘struct dp_netdev_flow’ has no member named ‘offloaded_actions_len’
                                &flow->offloaded_actions_len);

it is resolved in the next commit

File: lib/dpif-netdev.c [hunk 2]
```diff
@@ -6796,7 +6814,7 @@ dfc_processing(struct dp_netdev_pmd_thread *pmd,
 
         if ((*recirc_depth_get() == 0) &&
             dp_packet_has_flow_mark(packet, &mark)) {
-            flow = mark_to_flow_find(pmd, mark);
+           flow = process_partial_offload(pmd, port_no, mark, packet);
```
Comment (elibritstein) [inline lib/dpif-netdev.c]: space (indentation)

File: lib/netdev-provider.h [hunk 3]
```diff
@@ -866,6 +866,18 @@ struct netdev_class {
                     size_t actions_len, const ovs_u128 *ufid,
                     struct offload_info *info, struct dpif_flow_stats *);
 
+    /* Restore the flow/packet metadata state for continued processing
+     * in software.
+     * 'actions' are in netlink format.
+     * 'offloaded_actions_len' is populated according to the offloaded actions
+     * Return 0 if successful, otherwise returns a positive errno value.
+     *
```
Comment (elibritstein) [inline lib/netdev-provider.h]: redundant * line

File: lib/netdev.c [hunk 4]
```diff
@@ -2206,6 +2206,22 @@ netdev_flow_put(struct netdev *netdev, struct match *match,
...
+                             size_t *offloaded_actions_len)
+{
+    const struct netdev_class *class = netdev->netdev_class;
+
+    return (class->flow_restore_state
+            ? class->flow_restore_state(netdev, flow_mark, packet,
+                                        actions, actions_len,
+                                        offloaded_actions_len)
+            : EOPNOTSUPP);
+
```
Comment (elibritstein) [inline lib/netdev.c]: redundant space line

File: lib/dpif-netdev.c [hunk 5]
```diff
@@ -6732,6 +6732,24 @@ smc_lookup_batch(struct dp_netdev_pmd_thread *pmd,
     pmd_perf_update_counter(&pmd->perf_stats, PMD_STAT_SMC_HIT, n_smc_hit);
 }
 
+static inline struct dp_netdev_flow *
```
Comment (elibritstein) [inline lib/dpif-netdev.c]: why inline?

File: lib/dpif-netdev.c [hunk 6]
```diff
@@ -573,8 +573,13 @@ struct dp_netdev_flow {
     struct ovs_refcount ref_cnt;
 
     bool dead;
+
+    /* HW offload state
+     *
```
Comment (elibritstein) [inline lib/dpif-netdev.c]: why not single line comment?

File: lib/netdev-rte-offloads.c [hunk 7]
```diff
@@ -1100,6 +1100,32 @@ netdev_rte_offloads_validate_flow(const struct match *match, bool ct_offload,
 static int
 netdev_offloads_flow_del(const ovs_u128 *ufid);
 
+int
+netdev_rte_offloads_flow_restore(OVS_UNUSED struct netdev *netdev,
+                                     OVS_UNUSED uint32_t flow_mark ,
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: identation

File: lib/netdev-rte-offloads.c [hunk 8]
```diff
@@ -1100,6 +1100,32 @@ netdev_rte_offloads_validate_flow(const struct match *match, bool ct_offload,
...
+                                     OVS_UNUSED struct dp_packet *packet,
+                                     struct nlattr *actions, size_t actions_len,
+                                     size_t *offloaded_actions_len)
+{
+    unsigned int left;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: rev xmas

Comment (ozshlomo) [reply lib/netdev-rte-offloads.c]: Done

File: lib/netdev-rte-offloads.c (hunk 7 above)
Comment (ozshlomo) [reply lib/netdev-rte-offloads.c]: Done

File: lib/dpif-netdev.c (hunk 6 above)
Comment (ozshlomo) [reply lib/dpif-netdev.c]: Done

File: lib/netdev.c (hunk 4 above)
Comment (ozshlomo) [reply lib/netdev.c]: Done

File: lib/dpif-netdev.c [hunk 9]
```diff
@@ -6732,6 +6732,24 @@ smc_lookup_batch(struct dp_netdev_pmd_thread *pmd,
...
 
+static inline struct dp_netdev_flow *
+process_partial_offload(struct dp_netdev_pmd_thread *pmd, odp_port_t port_no,
+                           uint32_t mark, struct dp_packet *packet)
+{
+    struct dp_netdev_port *port = dp_netdev_lookup_port(pmd->dp, port_no);
+    struct dp_netdev_flow *flow = NULL;
+    struct dp_netdev_actions* actions = NULL;
+
+    flow = mark_to_flow_find(pmd, mark);
```
[the remark this answers is not in the document]
Comment (ozshlomo) [reply lib/dpif-netdev.c]: I disagree. The platform always call restore. It is up to the implementation to decide what to do.

File: lib/dpif-netdev.c [hunk 10]
```diff
@@ -6732,6 +6732,24 @@ smc_lookup_batch(struct dp_netdev_pmd_thread *pmd,
...
+    if (flow)
+        actions = dp_netdev_flow_get_actions(flow);
+
```
[the remark this answers is not in the document]
Comment (ozshlomo) [reply lib/dpif-netdev.c]: Already fixed

**Pull request merged**